README
-------------------------------------------------------------------

     =========================================================
     Geant4 - an Object-Oriented Toolkit for Simulation in HEP
     =========================================================

                            CalSG
                            -------
  




This program is based on TestEm3 and its commands can still be used.

Here 5 blocks can be created individually. The following parameters can be set via commands.

/testem/det/setBlock 0 -20 cm 0 cm 0 cm 2 cm 10.0 cm 10.0 cm Aluminium 1 -> Block number  X Y Z Position  X Y Z  Length  Material Number of copies

Number of blocks
Thickness, height and depth
Position
Material
Number of copies which are placed one after the other


With this command a block can be further modified.

/testem/det/setAbsorBlock 4  Lead -> Block number Material First layer

Whether a block should consist of two layers
Material of the layer in front in X direction


CalSG.cc also has a Python mode.
CalSG is started with Control_v1.py and the position can be changed with commands from the console while the programme is in idle mode.
iu -> Block i goes up one cm.
id -> Block i moves down one cm.
q -> Program is terminated.


With HepRepFile a HepRep file can be created which stores the geometry, positions of the tracks and the global time for each step.






From here on the documentation of TestEm3 follows.




 How to collect energy deposition in a sampling calorimeter.
 How to survey energy flow.
 how to print stopping power. 

 1- GEOMETRY DEFINITION
 
  The calorimeter is a box made of a given number of layers.
  A layer consists of a sequence of various absorbers (maximum MaxAbsor=9).
  The layer is replicated.
 
  Parameters defining the calorimeter :
    - the number of layers,
    - the number of absorbers within a layer,		
    - the material of the absorbers,
    - the thickness of the absorbers,
    - the transverse size of the calorimeter (the input face is a square). 
 
  In addition a transverse uniform magnetic field can be applied.
 
  The default geometry is constructed in DetectorConstruction class, but all
  of the above parameters can be modified interactively via the commands 
  defined in the DetectorMessenger class.


        |<----layer 0---------->|<----layer 1---------->|<----layer 2---------->|
        |           |           |                       |                       |
        ==========================================================================
        ||          |           ||          |           ||          |           ||
        ||          |           ||          |           ||          |           ||
        ||   abs 1  | abs 2     ||   abs 1  | abs 2     ||   abs 1  | abs 2     ||
        ||          |           ||          |           ||          |           ||
        ||          |           ||          |           ||          |           ||
 beam   ||          |           ||          |           ||          |           ||
======> ||          |           ||          |           ||          |           ||
        ||          |           ||          |           ||          |           ||
        ||          |           ||          |           ||          |           ||
        ||          |           ||          |           ||          |           ||
        ||          |           ||          |           ||          |           ||
        ||   cell 1 | cell 2    ||   cell 3 | cell 4    ||   cell 5 | cell 6    ||
        ==========================================================================
        ^           ^           ^           ^           ^           ^           ^
        pln1        pln2        pln3       pln4        pln5        pln6       pln7
 
  NB. The number of absorbers and the number of layers can be set to 1.
  In this case we have a unique homogeneous block of matter, which looks like 
  a bubble chamber rather than a calorimeter ...
  (see the macro emtutor.mac)
 
 2- PHYSICS LISTS
 
  Physics lists are based on modular design. Several modules are instantiated:
  1. Transportation
  2. EM physics
  3. Decays
  4. StepMax - for step limitation

  EM physics builders can be local (eg. in this example) or from G4 kernel
  physics_lists subdirectory.

  Local physics builders:	 
  - "local"	standard EM physics with current 'best' options setting.
              these options are explicited in PhysListEmStandard
  
  From geant4/source/physics_lists/builders:	
  - "emstandard_opt0" recommended standard EM physics for LHC
  - "emstandard_opt1" best CPU performance standard physics for LHC
  - "emstandard_opt2" similar fast simulation
  - "emstandard_opt3" best standard EM options - analog to "local" above
  - "emstandard_opt4" best current advanced EM options standard + lowenergy
  - "emstandardWVI" standard EM physics and WentzelVI multiple scattering
  - "emstandardSS"  standard EM physics and single scattering model
  - "emlivermore"  low-energy EM physics using Livermore data
  - "empenelope"   low-energy EM physics implementing Penelope models
  - "emlowenergy"  low-energy EM physics implementing experimental
                    low-energy models
      
  Physics lists and options can be (re)set with UI commands.
    
 3- AN EVENT : THE PRIMARY GENERATOR
 
  The primary kinematic consists of a single particle which hits the calorimeter
  perpendicular to the input face. The type of the particle and its energy are 
  set in the PrimaryGeneratorAction class, and can be changed via the 
  G4 build-in commands of G4ParticleGun class (see the macros provided with this 
  example).
 	
  In addition one can choose randomly the impact point of the incident particle.
  The corresponding interactive command is built in PrimaryGeneratorMessenger.
 	
  A RUN is a set of events.
  
  CalSG computes the energy deposited per absorber and the energy flow through
  the calorimeter.
 				
 4- VISUALIZATION
 
  The Visualization Manager is set in the main() (see CalSG.cc).
  The initialisation of the drawing is done via the commands :
  /vis/... in the macro vis.mac. In interactive session:
  PreInit or Idle > /control/execute vis.mac
 	
  The default view is a longitudinal view of the calorimeter.
 	
  The tracks are drawn at the end of event, and erased at the end of run.
  Optionally one can choose to draw all particles, only the charged ones, or
  none. This command is defined in EventActionMessenger class.
 	
 5- PHYSICS DEMO
 
  The particle's type and the physics processes which will be available
  in this example are set in PhysicsList class.
 	
  In addition a built-in interactive command (/process/inactivate processName)
  allows to activate/inactivate the processes one by one.
  Then one can well visualize the processes one by one, especially 
  in the bubble chamber setup with a transverse magnetic field.
 
  As a homework try to visualize a gamma conversion alone, 
  or the effect of the multiple scattering.

  Notice that one can control the maximum step size, via the
  StepMax process and the command /testem/stepMax
  (see StepMax and PhysicsList classes)
 
 6- HOW TO START ?
 
  - Execute CalSG in 'batch' mode from macro files
      % CalSG   run01.mac
 
  - Execute CalSG in 'interactive mode' with visualization
    % CalSG
    ....
    Idle> type your commands. For instance:
    Idle> /control/execute run01.mac
    ....
    Idle> exit

  Macros provided in this example:
  - atlashec.mac: ATLAS HEC model
  - dedx.mac: to control dE/dx calculation: 1 layer; minimum ionizing particle
  - emtutor.mac: for tutorial; interactivity + visualisation
  - geom.mac: to play with geometry
  - ionC12.mac: ion C12, 1 layer
  - lhcb.mac: LHCB ECAL model
  - linac.mac: Linac/Ecal from Graham Wilson
  - lockwood.mac: Al-Au-Al 1 layer (G.L.Lockwood et al. SAND79-0414 (1980))
  - run01.mac: Lead-liquidArgon 50 layers; electron 1 GeV
  - run02.mac: Tungsten-Silicon 50 layers; electron 1 GeV
  - storeTables.mac: show how to store and retrieve physics tables
  - tileCal.mac: ATLAS tileCal
  - vis.mac: to activate visualization

 7- HISTOGRAMS
 
 CalSG can produce histograms : 
  histo 1 : energy deposit in absorber 1
  histo 2 : energy deposit in absorber 2
  ...etc...........
    
  histo 11 : longitudinal profile of energy deposit in absorber 1 (MeV/event)
  histo 12 : longitudinal profile of energy deposit in absorber 2 (MeV/event)  
  ...etc...........  
  
  histo 21 : energy flow (MeV/event)
  histo 22 : lateral energy leak (MeV/event)  
  ...etc...........
      
  NB. Numbering scheme for histograms:
  layer     : from 1 to NbOfLayers (included)
  absorbers : from 1 to NbOfAbsor (included)
  planes    : from 1 to NbOfLayers*NbOfAbsor + 1 (included)
  
 One can control the binning of the histo with the command:
  /analysis/h1/set   idAbsor  nbin  Emin  Emax  unit 
  where unit is the desired energy unit for that histo (see CalSG.in).
         
  One can control the name of the histograms file with the command:
  /analysis/setFileName  name  (default CalSG)
   
  It is possible to choose the format of the histogram file : root (default),
  xml, csv, by using namespace in HistoManager.hh 
   	
 It is also possible to print selected histograms on an ascii file:
 /analysis/h1/setAscii id
 All selected histos will be written on a file name.ascii  (default CalSG)
    


History
----------------------------------------------------

     =========================================================
     Geant4 - an Object-Oriented Toolkit for Simulation in HEP
     =========================================================

                    CalSG History file
                    --------------------
This file should be used by the G4 example coordinator to briefly
summarize all major modifications introduced in the code and keep
track of all tags.

     ----------------------------------------------------------
     * Reverse chronological order (last date on top), please *
     ----------------------------------------------------------
     
06-10-21 I. Hrivnacova (CalSG-V10-07-03)
- Migration to new G4AnalysisManager.hh header;
  define the default output file type (root).

19-07-21 I. Hrivnacova (CalSG-V10-07-02)
- Updated for changes in the analysis category:
  removed deleting of the analysis manager,
  as this is now done by the Geant4 kernel.

06-05-21 mma (CalSG-V10-07-01)
- Migration to G4RunManagerFactory and G4SteppingVerboseWithUnits.
       
14-12-20 mma (CalSG-V10-07-00)
- updated PhysListEmStandard.cc
- StepMax, StepMaxMessenger, PhysicsList, PhysicsListMessenger :
    revert, to make command /testem/stepMax working in interactive mode
- added macro stepMax.mac
      
14-09-20 I. Hrivnacova (CalSG-V10-06-01)
- Renamed CalSG_2.in in geom.mac

01-09-20 mma, I. Hrivnacova (CalSG-V10-06-00)
- Macro & commands review:
  - Added macro descriptions in README
  - Added CalSG_2.in with commands not used in CalSG.in
  - Fixed not working macros: ionC12.mac, lockwood.mac, storeTables.mac
  - Removed obsolete macro retrieveTables.mac 

26-11-19 I. Hrivnacova (CalSG-V10-05-00)
- Fixed Doxygen warnings

09-05-18 B.Morgan (CalSG-V10-04-06)
- Include G4Types before use of G4MULTITHREADED. For forward
  compatibility with move to #defines over -D for G4 preprocessor
  symbols.

29-03-18 V.Ivant (CalSG-V10-04-05)
- CalSG.cc - small optimisation

27-03-18 mma (CalSG-V10-04-04)
- remove radioactive.mac

26-03-18 V.Ivant (CalSG-V10-04-03)
- DetectorConstruction - fixed initialisation

23-03-18 V.Ivant (CalSG-V10-04-02)
- StepMax - use thread safe instantiation, mesenger merged with 
    PhysicsListMessenger
- DetectorConstruction - improved method names, not allow to clear 
    geometry between runs
- PhysicsList - make Opt0 default, removed local instantiation of G4Decay
- Use MixMax default random number generator 

13-03-18 mma (CalSG-V10-04-01)
- testem1.cc : remove G4UI_USE and G4VIS_USE

14-12-17 V.Ivant (CalSG-V10-04-00)
- corrected CalSG.in to optimise output

14-10-16 G.Folger (CalSG-V10-02-06)
- remove direct use of theParticleIterator, use GetParticleTableIterator().
    fix required by clang39 on Linux and MAC

22-07-16 mma  (CalSG-V10-02-05)
- apply coding conventions

04-07-16 I. Hrivnacova  (CalSG-V10-02-04)
- Fixes for Doxygen documentation

10-02-16 mma (CalSG-V10-02-03)
- update radioactive.mac again

06-02-16 mma (CalSG-V10-02-02)
- update radioactive.mac

29-01-16 mma (CalSG-V10-02-01)
- Physicsist: add radioactiveDecay process (not active by default)

10-12-15 V.Ivant (CalSG-V10-02-00)
- H.Burkhardt fixed root macro allowing to work both with root5 and 6
    
28-10-15 D.Sawkey (CalSG-V10-01-04)
- update description of physics lists in README, again

26-10-15 D.Sawkey (CalSG-V10-01-03)
- update description of physics lists in README

19-09-15 V.Ivant (CalSG-V10-01-02)
- CalSG.cc - use default 4 threads, improve vis.mac
- PhysListEmStandard - use G4EmParameters

12-05-15 V.Ivant (CalSG-V10-01-01)
- PhysicsList - use G4EmStandardPhysicsGS from physics_list library

23-01-15 I. Hrivnacova (CalSG-V10-01-00)
- fixed formatting in .README
	 
28-10-14 mma (CalSG-V10-00-16)
- in macros : /run/numberOfThreads 2
 
23-10-14 V.Ivant (CalSG-V10-00-15)
- disable extra printout for sequential 
	 
23-10-14 V.Ivant (CalSG-V10-00-14)
- improved output MT 
	 
23-10-14 mma (CalSG-V10-00-13)
- PhysListEmStandard : msc algorithm fUseSafetyPlus

21-10-14 mma (CalSG-V10-00-12)
- PhysicsList : added G4EmStandardPhysicsSS
- CalSG.cc : option set nThreads via command
	 
27-08-14 mma (CalSG-V10-00-11)
- TestEm1.cc : SetNumberOfThreads( G4Threading::G4GetNumberOfCores() )

13-08-14 mma (CalSG-V10-00-10)
- update .README again

12-08-14 mma (CalSG-V10-00-09)
- update .README

11-08-14 V.Ivant (CalSG-V10-00-08)
- added "emstandardSS" option for EM physics 
	 
10-08-14 mma (CalSG-V10-00-07)
- update README, emtutor.mac
		 
14-06-14 mma (CalSG-V10-00-06)
- remove PhysicsList::SetCuts()	
 
11-06-14 mma (CalSG-V10-00-05)
- retag of 04
 
07-06-14 mma (CalSG-V10-00-04)
- minor cleanup in PhysicsList::SetCuts()
	 
05-06-14 mma (CalSG-V10-00-03)
- replace /testem/phys/setCuts by /run/setCut
 
28-04-14 V.Ivant (CalSG-V10-00-02)
- added "emlowenergy" option for EM physics 
     
14-01-14 mma (CalSG-V10-00-01)
- MT: Add forgotten classes ActionInitialization, Run
     
06-01-14 mma (CalSG-V10-00-00)
- MT: ActionInitialization, CalSG.cc
- MT: Run, RunAction, EventAction, TrackingAction, SteppingAction
- MT: DetectorConstruction and Messenger
- rm EventActionMessenger, hbook

24-07-13 P.Gumplinger (CalSG-V09-06-09)
- remove unused field fEventAct from TrackingAction class

11-07-13 I.Hrivnacova (CalSG-V09-06-08) 
- Updated CMakeLists.txt and SharedFilesList.txt for added classes
  in Hbook analysis code
- In HistoManager: use new analysis manager method SetH1Activation()   
    
19-06-13 mma (CalSG-V09-06-07)
- SteppingVerbose: use G4Step::GetSecondaryInCurrentStep() .
     
06-06-13 mma (CalSG-V09-06-06)
- Cosmetic in SteppingAction.

14-04-13 I.Hrivnacova (CalSG-V09-06-05)
- Removed ExG4HbookAnalysisManager and related files and updated CMake/GNUmake
  files to use it from extended/common          
     
01-04-13 A.Dotti (CalSG-V09-06-02,CalSG-V09-06-03,CalSG-V09-06-04)
- New G4VUserPhysicsList interface 
     
20-03-13 mma (CalSG-V09-06-01)
- PhysListEmStandard : use default G4UrbanMscModel

Control_v3.py
import os
import numpy as np
import time
import subprocess
#os.system('/build/make')
#os.system('./build/CalSG Run_Beam_v1.mac ')
import keyboard
import time

# get the start time
st = time.time()

##############################################
#Define Parameters
##############################################
BlockPosition = [[-20,-10,0,10,20],[0,0,0,0,0],[0,0,0,0,0]]
BlockSize = [[2,2,2,2,2],[10,10,10,10,10],[10,10,10,10,10]]
BlockPosition_D = [[-20,-10,0,10,20],[0,0,0,0,0],[0,0,0,0,0]]
#BlockSize_D = [[50,0,0,0,0],[10,10,10,10,10],[10,10,10,10,10]]
BlockSize_D = [[5,5,5,5,5],[20,20,20,20,20],[20,20,20,20,20]]
Material = ["Aluminium","Aluminium","Aluminium","Aluminium","Aluminium"]

print(BlockSize)
print(BlockSize[1][0])
##############################################
# Change Parameters
##############################################
def Check_Position(N,X,Y,Z):
    OXP = BlockPosition[0][N]
    OXS = BlockSize[0][N]
    for i in range(5):
        if i != N and np.abs(BlockPosition[0][i] - X) < (BlockSize[0][i]/2 + BlockSize[0][N]/2):
            #print( i , N , BlockPosition[0][N], X, np.abs(BlockPosition[0][N] - X) , (BlockSize[0][i]/2 + BlockSize[0][N]/2))
            #BlockSize[0][i] =  np.abs(BlockPosition[0][N] - X)/2
            BlockSize[0][N] =  np.abs(BlockPosition[0][N] - X) - BlockSize[0][N]


            if np.abs(BlockPosition[0][i] - X) <= BlockSize[0][i]/2:
                #BlockSize[0][i] =  0.0
                BlockSize[0][N] =  0.0

        if i != N and BlockSize[0][N] == 0.0  and np.abs(BlockPosition[0][i] - X) != 0.0:
            #BlockSize[0][i] =  BlockSize_D[0][i]
            BlockSize[0][N] =  BlockSize_D[0][N]

            if  np.abs(BlockPosition[0][i] - X) < (BlockSize[0][i]/2 + BlockSize[0][N]/2):
                #print( i , N , BlockPosition[0][N], X, np.abs(BlockPosition[0][N] - X) , (BlockSize[0][i]/2 + BlockSize[0][N]/2))
                #BlockSize[0][i] =  np.abs(BlockPosition[0][N] - X)/2
                BlockSize[0][N] =  np.abs(BlockPosition[0][N] - X) - BlockSize[0][N]

                if np.abs(BlockPosition[0][i] - X) <= BlockSize[0][i]/2:
                    #BlockSize[0][i] =  0.0
                    BlockSize[0][N] =  0.0
            
        

  
    return 0

def Check_Position_v2(N,X,Y,Z,B):
    BlockSizeN = B
    for i in N:
        for j in N:
            if i< j:
                BlockSizeN[0][i] =  B[0][i]
                BlockSizeN[0][j] =  B[0][j]
                #print( i, j)
                #print(BlockSizeN)
                #print( np.abs(X[i] - X[j]) , (BlockSizeN[0][i]/2 + BlockSizeN[0][j]/2))
                if i < j  and np.abs(X[i] - X[j]) < (BlockSizeN[0][i]/2 + BlockSizeN[0][j]/2):
                    BlockSizeN[0][i] =  np.abs(X[i] - X[j])/2#*0.001
                    BlockSizeN[0][j] =  np.abs(X[i] - X[j])/2#*0.001

                    print( i, j)

                    if  np.abs(X[i] - X[j]) == 0.0:
                        BlockSizeN[0][i] =  0.0
                        BlockSizeN[0][j] =  0.0
                        print( i, j)

  

             
            if i < j and np.abs(X[i] - X[j]) != 0.0 and BlockSizeN[0][i] == 0.0 and BlockSizeN[0][j] == 0.0 :
                BlockSizeN[0][i] =  BlockSize_D[0][i]
                BlockSizeN[0][j] =  BlockSize_D[0][j]
                print( i, j)

                if np.abs(X[i] - X[j]) < (BlockSizeN[0][i]/2 + BlockSizeN[0][j]/2):
                    BlockSizeN[0][i] =  np.abs(X[i] - X[j])/2#*0.001
                    BlockSizeN[0][j] =  np.abs(X[i] - X[j])/2#*0.001
                    print( i, j)
            if i != j and X[i] == X[j]:
                BlockSizeN[0][i] =  0.0
                BlockSizeN[0][j] =  0.0
                print( i, j)
           # BlockSize = BlockSizeN
        
    return BlockSizeN








def SetBlockPosition(N,X,Y,Z,M):
    Check_Position(N,X,Y,Z)

    Text = "/testem/det/setBlock " + str(N) + " " + str(X) + " cm "+ str(Y) + " cm "+ str(Z) + " cm " + str(BlockSize[0][N]) + " cm 10.0 cm 10.0 cm " + M + " 1 \n"
    a_file = open('build/Run_Beam_v1.mac', "r")
    list_of_lines = a_file.readlines()
   # print(list_of_lines)
    a_file = open('build/Run_Beam_v1.mac', "w")
    list_of_lines[14 + N] = Text
    a_file.writelines(list_of_lines)
    a_file.close()
    return 0



def SetBlockPosition_v2(N,X,Y,Z,M,B):
    BlockN = Check_Position_v2(N,X,Y,Z,B)

    a_file = open('build/Run_Beam_v1.mac', "r")
    list_of_lines = a_file.readlines()
    a_file = open('build/Run_Beam_v1.mac', "w")

    for i in N:
        Text = "/testem/det/setBlock " + str(i) + " " + str(X[i]) + " cm "+ str(Y[i]) + " cm "+ str(Z[i]) + " cm " + str(BlockN[0][i]) + " cm 10.0 cm 10.0 cm "+ M[i] + " 1 \n"
        list_of_lines[14 + i] = Text
        print(i, Text)
    a_file.writelines(list_of_lines)

    a_file.close()
    return 0


def SetBlockPosition_v3(N,X):
    #BlockN = Check_Position_v2(N,X,Y,Z,B)

    a_file = open('build/Run_Beam_v2.mac', "r")
    list_of_lines = a_file.readlines()
    a_file = open('build/Run_Beam_v2.mac', "w")

    
    TX = list_of_lines[14 + N]
    print(TX)
    print(len(TX))
    Xi = TX[22+8:26+5]
    print(Xi)

    Text = TX[0:30] + str(int( Xi) + X[N] ) + TX[31:74]

    list_of_lines[14 + N] = Text
    print(N, Text)
    a_file.writelines(list_of_lines)

    a_file.close()
    return 0

#import subprocess

#DCalSG = "/home/kappe/projects/CERN_SS/Geant_Project/CalSG_1Block_moveble/build/CalSG"
#proc = subprocess.Popen("/home/kappe/projects/CERN_SS/Geant_Project/CalSG_1Block_moveble/build/CalSG",




def Cpp_Execution(W):
    proc = subprocess.Popen(["./build/CalSG", "Python"],
    stdin=None ,#subprocess.PIPE,
    stdout=subprocess.PIPE,
    universal_newlines=False)

   # inp = input("Message to CPP>> ")

   # proc.stdin.write("/run/beamOn 1")
   # stdout, stderr = proc.communicate()
    #proc.wait()
   # print(stdout)


    #cppMessage = proc.stdout.readline().rstrip("\n") 
    #print( "cppreturn message ->" + cppMessage + " written by python \n")
    

    # creating a pipe to child process

    data, temp = os.pipe()

    # writing inputs to stdin and using utf-8 to convert it to byte string
    if (W):
        os.write(temp, bytes("run", "utf-8"));

        os.close(temp)
    if (W == False):
        os.write(temp, bytes("quit", "utf-8"));

        os.close(temp)


    subprocess.call(["./build/CalSG"],stdin=data, stdout=None, stderr=None, shell=False)





NChanges = 0
#os.system('./build/CalSG Run_Beam_v1.mac ')
M3 = ["Lead","Lead","Lead","Lead","Lead","Lead","Aluminium","Aluminium","Aluminium","Aluminium",]
'''
for i in range(10):
    M = ["Aluminium","Scintillator","Aluminium","Aluminium",M3[i]]
    N = [0,1,2,3,4]
    X = [-20 + i*4,-10 - i, 0, 10,20]
    X = [-30+i*2,-10 - i, 0, 10,20]
    Y = [i*2,i,0,-1,i]
    Z = [0,0,0,0,0]
    BlockSize_D = [[2,2,2,2,2],[20,20,20,20,20],[20,20,20,20,20]]

   # SetBlockPosition_v2(N,X,Y,Z,M,BlockSize_D)
    NChanges = NChanges + 1
'''
q = True


proc = subprocess.Popen(["./CalSG", "Python"],
stdin=None ,#subprocess.PIPE,
stdout=subprocess.PIPE)#,
universal_newlines=False)

data, temp = os.pipe()
#subprocess.call("source /home/kappe/Geant4_11_0_2/bin/geant4.sh ",stdin=None,stdout=None,stderr=None,shell=False)
while(q):
    value = input("Please enter an integer:\n")
    print(value)
    if keyboard.is_pressed("1"):
        X = [1,0,0,0,0]
        # SetBlockPosition_v3(0,X)
        #os.write(temp, bytes("run", "utf-8"))
        proc.communicate()
       # os.close(temp)
      #  subprocess.call(["./build/CalSG"],stdin=data, stdout=None, stderr=None, shell=False)
        proc.wait(timeout=None)

    if keyboard.is_pressed("2"):
        X = [-1,0,0,0,0]
        # SetBlockPosition_v3(0,X)
        os.write(temp, bytes("run", "utf-8"))
      #  os.close(temp)
       # subprocess.call(["./build/CalSG"],stdin=data, stdout=None, stderr=None, shell=False)
        proc.wait(timeout=None)
    if keyboard.is_pressed("q"):
        os.write(temp, bytes("quit", "utf-8"))
      #  os.close(temp)
       # subprocess.call(["./build/CalSG"],stdin=data, stdout=None, stderr=None, shell=False)
        proc.wait(timeout=None)
        q = False
  
    

et = time.time()

# get the execution time
elapsed_time = et - st
print('Execution time:', elapsed_time/NChanges, 'seconds', "\n" , "For mean time 10 changes with 1 events each")
